core/arch/arm/sm/pm_a32.S

```diff
@@ -0,0 +1,223 @@
+	pop	{r0, pc}
+aborted:
+	/* cpu not suspended */
+	add	sp, sp, #SM_PM_CTX_SIZE
+	/* Return -1 to the caller */
+	mov	r0, #(-1)
+suspend_return:
+	pop	{r4 - r12, pc}
+UNWIND( .fnend)
+END_FUNC sm_pm_cpu_suspend
+
+FUNC sm_pm_cpu_do_suspend, :
+UNWIND(.fnstart)
+UNWIND(.cantunwind)
+	push	{r4 - r11}
+	read_midr r4
+	ubfx	r5, r4, #4, #12
+	ldr     r4, =CORTEX_A7_PART_NUM
+	cmp	r5, r4
+	beq	a7_suspend
+	ldr     r4, =CORTEX_A9_PART_NUM
+	cmp	r5, r4
+	beq	a9_suspend
+	/* cpu not supported */
+	b	.
+	/* A9 needs PCR/DIAG */
+a9_suspend:
+	read_pcr r4
+	read_diag r5
+	stmia	r0!, {r4 - r5}
+a7_suspend:
+	read_fcseidr r4
+	read_tpidruro r5
+	stmia	r0!, {r4 - r5}
+	read_dacr r4
+#ifdef CFG_WITH_LPAE
+#error "Not supported"
+#else
+	read_ttbr0 r5
+	read_ttbr1 r6
+	read_ttbcr r7
+#endif
+	read_sctlr r8
+	read_actlr r9
+	read_cpacr r10
+	read_mvbar r11
+	stmia	r0!, {r4 - r11}
+	read_prrr r4
+	read_nmrr r5
+	read_vbar r6
+	read_nsacr r7
+	stmia	r0, {r4 - r7}
+	pop	{r4 - r11}
+	bx	lr
+UNWIND( .fnend)
+END_FUNC sm_pm_cpu_do_suspend
+
+FUNC sm_pm_cpu_resume, :
+UNWIND(.fnstart)
+UNWIND( .cantunwind)
+	cpsid	aif
+
+	/* Call into get_core_pos */
+	adr	r0, _core_pos
+	ldr	r1, [r0]
+	add	r0, r0, r1
+	blx	r0
+
+	/*
+	 * At this point, MMU is not enabled now.
+	 * 1. Get the runtime physical address of _suspend_sp
+	 * 2. Get the offset from _suspend_sp to &thread_core_local
+	 * 3. Get the runtime physical address of thread_core_local
+	 * Since moving towards non-linear mapping,
+	 * `ldr r0, =thread_core_local` is not used here.
+	 */
+	adr	r4, _suspend_sp
+	ldr	r5, [r4]
+	add	r4, r4, r5
+
+	mov_imm r1, THREAD_CORE_LOCAL_SIZE
+	mla	r0, r0, r1, r4
+
+	ldr	r0, [r0, #THREAD_CORE_LOCAL_SM_PM_CTX_PHYS]
+	/* Need to use r0!, because sm_pm_cpu_do_resume needs it */
+	ldmia	r0!, {sp, pc}
+UNWIND( .fnend)
+END_FUNC sm_pm_cpu_resume
+
+/*
+ * void sm_do_cpu_do_resume(paddr suspend_regs) __noreturn;
+ * Restore the registers stored when sm_pm_cpu_do_suspend
+ * r0 points to the physical base address of the suspend_regs
+ * field of struct sm_pm_ctx.
+ */
+FUNC sm_pm_cpu_do_resume, :
+UNWIND(.fnstart)
+UNWIND( .cantunwind)
+	read_midr r4
+	ubfx	r5, r4, #4, #12
+	ldr     r4, =CORTEX_A7_PART_NUM
+	cmp	r5, r4
+	beq	a7_resume
+
+	/*
+	 * A9 needs PCR/DIAG
+	 */
+	ldmia   r0!, {r4 - r5}
+	write_pcr r4
+	write_diag r5
+
+a7_resume:
+	/* v7 resume */
+	mov	ip, #0
+	/* Invalidate icache to PoU */
+	write_iciallu
+	/* set reserved context */
+	write_contextidr ip
+	ldmia	r0!, {r4 - r5}
+	write_fcseidr r4
+	write_tpidruro r5
+	ldmia	r0!, {r4 - r11}
+	/* Invalidate entire TLB */
+	write_tlbiall
+	write_dacr r4
+#ifdef CFG_WITH_LPAE
+#error "Not supported -"
+#else
+	write_ttbr0 r5
+	write_ttbr1 r6
+	write_ttbcr r7
+#endif
+
+	ldmia	r0, {r4 - r7}
+	write_prrr r4
+	write_nmrr r5
+	write_vbar r6
+	write_nsacr r7
+
+	write_actlr r9
+	write_cpacr r10
+	write_mvbar r11
+	isb
+	dsb
+	/* MMU will be enabled here */
+	write_sctlr r8
+	isb
+	mov	r0, #0
+	b	suspend_return
```

Comment:
After reviewing these patches the only functions that are generic enough to be integrated into our(TI) suspend/resume implementation are 'sm_pm_cpu_do_suspend' and 'sm_pm_cpu_do_resume'.

One change we did have to make is this last line needed to be changed, we have to call into this function from another spot and so our stack is not in the same state as yours. If possible this function should be behave as a normal function (end with bx lr) so we can call it from our context.

If this patch could be isolated to just those functions I would like to take that and post our suspend/resume solution on top of that so we can compare/contrast and hopefully find more common ground.

Reply:
@glneo Per @jenswi-linaro 's comments, I changed from pop {r4-r12, pc} to b suspend_return. Why would you split the functions into stanalone files? Could you share you code or example? Then we could see how to proceed. Before I continue, I would hear voice from @jenswi-linaro and @etienne-lms

Reply:
We've been reviewing this for quite some time now and both me and @etienne-lms have no or only minor remarks left. Given that I'd rather get this merged soon.
After that we can continue and see how we can fit TI's suspend/resume into this. @glneo, I hope that's OK with you too.

Reply:
agree

Reply:
I have no specific issue with this being pulled. My comment above was made assuming this whole set would not be merged for some time, with that it would have worked better for us to split out this patch and merge first so we could base our work on it, if the whole set is about to get taken then that would work also.

Reply:
Thanks. I'll send out a new version to adrress @etienne-lms 's comments soon